catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryReportActionProvider.java

```diff
@@ -0,0 +1,224 @@
+import ddf.action.ActionProvider;
+import ddf.action.impl.ActionImpl;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.source.Source;
+
+public class RegistryReportActionProvider implements ActionProvider {
+
+    private static final String REGISTRY_PATH = "/registries";
+
+    private static final String REPORT_PATH = "/report";
+
+    private static final String FORMAT = ".html";
+
+    private static final String SOURCE_ID_QUERY_PARAM = "?sourceId=";
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(RegistryReportActionProvider.class);
+
+    private final String actionProviderId;
+
+    private ConfigurationAdmin configurationAdmin;
+
+    private String description;
+
+    private String title;
+
+    public RegistryReportActionProvider(String id) {
+        this.actionProviderId = id;
+    }
+
+    public <T> List<Action> getActions(T subject) {
+
+        if (canHandle(subject)) {
+            if (subject instanceof Metacard) {
+                return processSubject((Metacard) subject);
+            } else if (subject instanceof Source) {
+                return processSubject((Source) subject);
+            } else if (subject instanceof Configuration) {
+                return processSubject((Configuration) subject);
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    private List<Action> processSubject(Metacard metacard) {
+
+        if (metacard.getAttribute(RegistryObjectMetacardType.REGISTRY_ID) == null
+                || StringUtils.isBlank(metacard.getAttribute(RegistryObjectMetacardType.REGISTRY_ID)
+                .getValue()
+                .toString())) {
+            LOGGER.info("No registry id given. No action to provide.");
+            return Collections.emptyList();
+        }
+
+        try {
+            String registryId =
+                    URLEncoder.encode(metacard.getAttribute(RegistryObjectMetacardType.REGISTRY_ID)
+                            .getValue()
+                            .toString(), (StandardCharsets.UTF_8).toString());
+            String sourceId = URLEncoder.encode(getSource(metacard),
+                    (StandardCharsets.UTF_8).toString());
+            Action action = getAction(registryId, sourceId);
+
+            if (action == null) {
+                return Collections.emptyList();
+            }
+
+            return Collections.singletonList(action);
+
+        } catch (UnsupportedEncodingException e) {
+            LOGGER.info("Unsupported Encoding exception", e);
+            return Collections.emptyList();
+        }
+    }
+
+    private List<Action> processSubject(Source source) {
+        String configId = source.getId();
+        if (configId != null) {
+            Configuration[] configurations;
+            try {
+
+                configurations = configurationAdmin.listConfigurations(String.format("(id=%s)",
+                        configId));
+
+                if (configurations.length > 0) {
```

Comment:
Could also try something (without the length check, even) like...
```java
Arrays.stream(configurations)
        .map(Configuration::getProperties)
        .map(p -> p.get(RegistryObjectMetacardType.REGISTRY_ID))
        .filter(Objects::nonNull)
        .findFirst()
        .ifPresent(p -> processSubject(configurations[0]));
```
I know it's a little longer - owing to the chained `map` calls instead of doing them in a single operation - but is much more declarative.